### PR TITLE
[IMP] web_editor, *: add web_editor.media model

### PR DESCRIPTION
**Pull request by robinlej**
COMMIT 1
[MOV] web_editor: move image processing variables to utils
This commit moves some image processing variables to the utils file, in
order to be able to use them in a frontend bundle (specifically
`web_editor.assets_media_dialog`) in a further commit.

COMMIT 2
[IMP] web_editor: remove media dialog's optimized images toggle
It was decided to remove the possibility to show optimized images
directly in the media dialog, be it with the toggle in debug mode or
when clicking directly on an optimized image without debug mode.
The latter was already not working anymore, and the former is considered
not critical.
Those images are still visible in the "Attachments" view in the
settings.

The main reason behind this decision is to prepare for the next commits,
which will refactor the way attachments are uploaded through media
dialogs by using a `web_editor.media` model. In addition to a quick gain
in code complexity, removing this feature will allow to make a less
convoluted SQL query when fetching media records.

COMMIT 3
[IMP] web_editor, *: add web_editor.media model
*: website, web_unsplash, test_website, website_blog, website_forum,
website_sale

This commit creates a `web_editor.media` model in order to have more
possibilities moving forward with media uploaded through the media
dialog in the frontend.
Indeed, these objects would benefit from specific fields that do not
need to be in the more general `ir.attachment` model (e.g. "favorite"
to have a user-friendly way to quickly find an often used media,
"hidden" to keep it in the DB but hide it from the dialog, "tags",
etc.).

On creation, the media model is automatically linked to an
`ir.attachment` through a `Binary` field if possible. If the attachment
is not a binary, the attachment is created just after the media and
attached to it through the `res_id` (=`media.id`) and `res_model`
(=`'web_editor.media'`) fields. The "true" `res_model` and `res_id` are
thus stored in the media record.
Other than that, the media model should welcome media dialog-specific
fields in the future.

The structure is such that a newly uploaded media creates a `media`
record, which generates an `ir.attachment`. Upon save in the frontend,
the attachment may be optimized (specifically for images): those copies
are `ir.attachment` only tied to their parent `ir.attachment` (through
the `original_id` field); they bear no link to the original `media`
record.

COMMIT 4
[IMP] website, web_editor: remove deleted images from website
When removing an image through the media dialog, the attachment is
deleted. But if this image is used on the website, it is not removed
from the page being currently edited. This lead to errors and
unexpected behaviors if the user tried to apply an effect (filter /
shape) on it before saving. Commit [1] fixed the most pressing issues in
stable, but some remain. (See steps to reproduce below for more
details.)

This commit changes the behavior with a checkbox on the remove dialog
asking the user whether they want to also remove the image from their
website.
If the user doesn't check the checkbox (default): the image is not
deleted from the website, and is therefore not deleted from the DB. It
is only hidden from the media dialog.
If the user checks the checkbox: the image is removed from the DB and
from the current page being edited (including all optimized images
derivating from the deleted one). Images are replaced by the "not found"
placeholder image.

Steps to reproduce:
- Drop a "Text - Image" snippet
- Replace the image with an uploaded image
- Save
- Edit
- Open the media dialog
- Delete the image
- Close the media dialog
- Apply a filter or a shape on the image
- Save
=> Silently fails
- Save again
=> Image is in base64 inside the page

[1]: 472ea5f

task-3452267

### CODE REVIEW

File: addons/website/static/src/components/media_dialog/image_selector.js [hunk 1]
```diff
@@ -1,13 +1,106 @@
...
+import { AttachmentError } from "@web_editor/components/media_dialog/file_selector";
+import { markup } from "@odoo/owl";
+
+patch(AutoResizeImage.prototype, "image_attachment_website", {
+    /**
+     * @override
+     */
+    async remove() {
+        const ids = await this.props.getOriginalAndCopiesIds(this.props.id);
+        const dialogBody = markup(sprintf(`%s
```
Comment (robinlej) [inline addons/website/static/src/components/media_dialog/image_selector.js:17]: The dialog body takes a string passed to a `t-out` in the template. `markup` is thus the only possibility to add a checkbox there.

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
Thanks for this PR. Here is a first review.

In the commit message:
- the user whether they want => the users whether they want OR the user whether he

File: odoo/addons/base/models/ir_attachment.py [hunk 2]
```diff
@@ -532,7 +532,7 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
         discard_binary_fields_attachments = False
         if not any(arg[0] in ('id', 'res_field') for arg in args):
             discard_binary_fields_attachments = True
-            args.insert(0, ('res_field', '=', False))
+            args.insert(0, ('res_field', 'in', [False, 'hidden']))
```
[the remark this answers is not in the document]
Comment (robinlej) [reply odoo/addons/base/models/ir_attachment.py:535]: Attachments with a specified `res_field` don't appear in the attachments view. So to still be able to see the images that have been removed from the media dialog but not from the website (and so not from the DB), this is the simplest and least intrusive way I found.

File: addons/web_editor/static/src/components/media_dialog/file_selector.js [hunk 3]
```diff
@@ -274,6 +274,28 @@ export class FileSelector extends Component {
...
 
+    async getOriginalAndCopiesIds(attachmentId) {
+        // Optimized copies for attachments other than the one that was clicked
+        // to open the media selector are not fetched by default as they are
+        // not needed. This selects all the copies of the file to delete.
+        const attachmentsDomain = this.attachmentsDomain.toSpliced(
+            this.attachmentsDomain.findIndex(domain => domain[0] === "original_id" && domain[1] === "in"),
+            1,
+            "|", ["original_id", "=", attachmentId], ["id", "=", attachmentId]
+        );
```
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/static/src/components/media_dialog/file_selector.js:285]: >  not 100% sure we can already use toSpliced()

`toSpliced()` doesn't trigger any linting error at least, but you're right, it's probably a bit too early for a few % around the globe. I'll change that. :)

> This part is not easy to read IMO

I'll add a comment to be clearer.

File: odoo/addons/base/models/ir_attachment.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Guillaume-gdi) [reply odoo/addons/base/models/ir_attachment.py:535]: Maybe worth considering to add a new field on the attachment model then. In the Website override of course so that we don't touch `base`. Or is there any issue with that ?

**Comment by rdeodoo** [comment]
As discussed, this will be retarget to master (hopefully the PR from GDI will be enough to not kill the editor when the original image is deleted and you click on an img) with the use of a new web_editor.media table where we will be able to use proper fields to do what we try here (visible boolean field directly on the table instead of playing with attachment res_id=-1 or res_field="hidden" etc)

File: addons/web_editor/models/ir_attachment.py [hunk 4]
```diff
@@ -71,6 +71,15 @@ def _compute_image_size(self):
...
 
+    def _get_serve_attachment(self, url, extra_domain=None, order=None):
+        # Bypass the default IrAttachment._search() behavior which adds
+        # ('res_field', '=', False) to the domain if no `res_field` nor `id` is
+        # set in the domain, as we at least want to fetch ('res_field', 'in',
+        # (False, 'binary_content')). We add a meaningless id domain in case
+        # some other module would have a legit 'res_field' query.
+        extra_domain = (extra_domain or []) + [('id', '!=', False)]
+        return super()._get_serve_attachment(url, extra_domain, order)
+
```
Comment (bso-odoo) [inline addons/web_editor/models/ir_attachment.py:82]: Beware, this is not classic class inheritance ! This change will apply to all `ir.attachment` as soon as `web_editor` is installed... (aka always) And I doubt the other applications would expect this behavior to change.
Maybe try to condition this with a context key:
From caller:
```py
attachment.with_context(any_attachment=True).something()
```
In `_get_serve_attachment`:
```py
if self.env.context.get('any_attachment'):
    extra_domain = (extra_domain or []) + [('id', '!=', False)]
```

File: addons/web_editor/static/src/components/media_dialog/file_selector.js [hunk 5]
```diff
@@ -232,11 +229,21 @@ export class FileSelector extends Component {
...
+
     get attachmentsDomain() {
         const domain = [
             '&',
-            ['res_model', '=', this.props.resModel],
-            ['res_id', '=', this.props.resId || 0]
+            ['res_model', '=', 'web_editor.media'],
+            '|', '&', ['res_field', '=', 'binary_content'], ['type', '=', 'binary'],
+                ['type', '=', 'url'],
         ];
```
Comment (bso-odoo) [inline addons/web_editor/static/src/components/media_dialog/file_selector.js:247]: Just wondering, is there no Javascript equivalent of the `from odoo.osv.expression import AND, OR` tools ?

**Review by bso-odoo — COMMENTED** [review]
I have not looked at the UI yet, but here are a few comments on the python-side.

File: addons/web_editor/static/src/components/media_dialog/file_selector.js (hunk 5 above)
Comment (Guillaume-gdi) [reply addons/web_editor/static/src/components/media_dialog/file_selector.js:247]: Maybe `import { Domain } from "@web/core/domain";`

**Comment by bso-odoo** [comment]
I tried reproducing the initial "delete image" then "apply shape" bug - confirming the deletion without removing the image from the page. I got the following in my DB:
```sql
# select * from web_editor_media;
 id | res_id | create_uid | write_uid | res_model  | url | hidden |        create_date         |         write_date         
----+--------+------------+-----------+------------+-----+--------+----------------------------+----------------------------
  1 |      0 |          2 |         2 | ir.ui.view |     | t      | 2024-03-14 07:53:19.863545 | 2024-03-14 07:53:39.929364
(1 row)
# select id, name, mimetype, res_model, res_id, res_field from ir_attachment where id=328 or original_id=328;
 id  |         name          |   mimetype    |    res_model     | res_id |   res_field    
-----+-----------------------+---------------+------------------+--------+----------------
 328 | chocobointherain.jpeg | image/jpeg    | web_editor.media |      1 | binary_content
 329 | chocobointherain.webp | image/webp    | ir.ui.view       |      0 | binary_content
 330 | chocobointherain.svg  | image/svg+xml | ir.ui.view       |      0 | binary_content
(3 rows)
```
I think the `ir.ui.view`-related attachments should not have a reference to the `binary_content` field.

**Comment by bso-odoo** [comment]
The fix does not work outside of website.
Steps to reproduce:
- Go to user preferences.
- Edit "Email Signature".
- Insert an image with "/image".
- Upload image.
- Save.
- Go to user preferences again.
- Select image in "Email Signature".
- Click on "Replace" on the toolbar.
- Delete image.
=> The checkbox mentions website.
=> Upon confirmation, a traceback popup is displayed.

**Comment by bso-odoo** [comment]
It would be nice to have a screen to browse the new media records (list/tree view, form view, filters etc...), maybe nearby "Attachments" in the debug mode's "Settings/Technical" menu ?
On such a screen, I guess it would be nice to have a navigation facility towards the attachment that relates to it ? (maybe that's implicit if the `binary_content` field is on the form)

**Comment by robinlej** [comment]
Thanks for the check!

> It would be nice to have a screen to browse the new media records (list/tree view, form view, filters etc...), maybe nearby "Attachments" in the debug mode's "Settings/Technical" menu ?

The plan is to show them on the attachments page, I just haven't done it yet. (The problem right now being that the page filters out attachments with a `res_field` (you can find them if you explicitly search for `'res_field', '=', 'binary_content'`). Same problem everywhere. :sweat_smile: )

> I think the ir.ui.view-related attachments should not have a reference to the binary_content field.

Hmm... Why do you think it's necessary? (Not saying it's not, I just don't see an obvious reason.)
If so, should I erase all `res_field`s when copying an image (see `def modify_image()`)? Or add a specific if statement for `binary_content` only? (not sure a legit flow can lead to copied images with a `res_field`, which need to keep their reference to the `res_field`...)
Note that this would make up for the previous issue, as those copies would appear in the attachments page by default. (Which might be a good enough reason in itself...)

**Comment by bso-odoo** [comment]
> > I think the ir.ui.view-related attachments should not have a reference to the binary_content field.
> Hmm... Why do you think it's necessary?

As far as I understood (I might be wrong), the `res_field` is supposed to be a field of the `res_model` that is referenced... if the referenced model is a `web_editor.media` then such `binary_content` field exists, but it is not the case of `ir.ui.view`.

File: addons/web_editor/models/ir_attachment.py (hunk 4 above)
Comment (robinlej) [reply addons/web_editor/models/ir_attachment.py:82]: It's a reasonable idea, but I don't see how to apply it. We get there through `_serve_fallback`, `_serve_db`... Meaning that adding the context to those in `web_editor` (or even in `web_unsplash`, because I think it's only a problem to serve unsplash images) poses the same issue.
Basically, this covers the `/unsplash/<unsplash_id>/<name>.<ext>` route, which isn't served by a dedicated route.

File: addons/web_editor/static/src/components/media_dialog/file_selector.js (hunk 5 above)
Comment (robinlej) [reply addons/web_editor/static/src/components/media_dialog/file_selector.js:247]: I don't understand when/how to use it, especially when all the fields are spelled out like here. TBH, it's a pain to write  and makes it less readable in my view. (The parameter of `Domain.and()` and `Domain.or()` must be an array, which itself contains full domains (so nested arrays), which are either composed by just one array or several arrays. And in the end I still need to cast it to a list anyway, for the logic of the other modules to work. Am I missing something? I must be missing something, it doesn't seems right.)

```js
get attachmentsDomain() {
    let domain = Domain.and([
        [['res_model', '=', 'web_editor.media']],
        Domain.or([
            [["res_field", "=", "binary_content"], ["type", "=", "binary"]],
            [['type', '=', 'url']],
        ]),
    ]);
    domain = Domain.or([[['public', '=', true]], domain]);
    domain = Domain.and([domain, [['name', 'ilike', this.state.needle]]]);
    return domain.toList();
}
```

File: addons/web_editor/models/ir_attachment.py (hunk 4 above)
Comment (bso-odoo) [reply addons/web_editor/models/ir_attachment.py:82]: Then maybe include this method in `web_unsplash`, and use the request's path in the condition (e.g. match id against a regex) ?

File: addons/web_editor/static/src/components/media_dialog/file_selector.js (hunk 5 above)
Comment (bso-odoo) [reply addons/web_editor/static/src/components/media_dialog/file_selector.js:247]: It's a bit less worse if you import `or` and `and`.
Not great indeed, but at least it's a way to get rid of the `push` and the `unshift`... and it makes the tree structure visible (at least for me - but I tend to see everything as trees, so I might be non-representative of the general audience for this code).
```js
return and([
    [['name', 'ilike', this.state.needle]],
    or([
        [['public', '=', true]],
        and([
            [['res_model', '=', 'web_editor.media']],
            or([
                [['res_field', '=', 'binary_content'], ['type', '=', 'binary']],
                [['type', '=', 'url']],
            ]),
        ]),
    ]),
]).toList();
```

Comment (robinlej) [reply addons/web_editor/static/src/components/media_dialog/file_selector.js:247]: > It's a bit less worse if you import or and and.

You can't import a class' static method.
I think I'll leave it as it is because all the other uses of `attachmentsDomain` do `push` and `unshift`, so let's keep the same writing "convention" in all of them. If further reviewers agree with your suggestion, I'll change it.

**Comment by robinlej** [comment]
@qsm-odoo Now that the upgrade PR passed and is green, everything is ready for a review. :crossed_fingers:

**Comment by robinlej** [comment]
About the 3 last pushes: 1) rebase + resolving a conflict, 2) style, 3) failing tests after resolving the conflicts, fixing what failed.

File: addons/web_editor/controllers/main.py [hunk 6]
```diff
@@ -281,15 +286,51 @@ def add_data(self, name, data, is_image, quality=0, width=0, height=0, res_id=Fa
...
+        media = self._media_create(
+            name=preprocessed_data['name'],
+            data=preprocessed_data['data'],
+            mimetype=preprocessed_data.get('mimetype'),
+            res_id=res_id,
+            res_model=res_model
+        )
+        return media._get_media_info()
+
+    @http.route('/web_editor/attachment/add_data', type='json', auth='user', methods=['POST'], website=True)
```
Comment (qsm-odoo) [inline addons/web_editor/controllers/main.py:312]: Mmh, I did not expect we would have to keep this one for two use cases, maybe we should not:

1. Media dialog "Files" -> Don't you think this should use the new `web_editor.media` model too? Maybe with a field `type` to distinguish Images, Files and maybe later Videos ? When we'll implement the favorite and/or tags stuff, we'll probably want the same for files, no? (even if it is later move as a link dialog feature rather than a media dialog feature). 

2. Logo upload during website configurator flow, can't it be also a new `web_editor.media` record. Really, the logo should always be suggested in the media dialog whatever page or record you are.

What do you think?

Also cc @rdeodoo in case he does not agree :)

File: addons/web_editor/models/media.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            # To avoid useless queries, we only fetch them if we have an image.
+            if attachment['mimetype'].split('/')[0] == 'image':
+                image_info = IrAttachment.search_read(
+                    [('id', '=', attachment['id'])],
+                    ['image_src', 'image_width', 'image_height'],
+                )[0]
+                attachment.update(image_info)
+            attachments.append(attachment)
+
+        return attachments
```
Comment (qsm-odoo) [inline addons/web_editor/models/media.py:129]: I was not expecting a SQL query when making those specs:
- What justify it? And if it really is justified, we should add a comment about it in the docstring
- Why can't it just be a `search_read` done client-side (so no python code for this), which returns the bunch of web_editor.media ids you need? Do you actually need anything but the ids of the media when opening the media dialog?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some questions before reviewing further and deeper :pray:

File: addons/web_editor/static/src/components/media_dialog/image_selector.js [hunk 8]
```diff
@@ -128,31 +126,6 @@ export class ImageSelector extends FileSelector {
...
-            // selected image when opening the media dialog.
-            // We might get a few more optimized image than necessary if the
-            // original image has multiple optimized images but it's not a
-            // big deal.
-            const originalId = this.props.media && this.props.media.dataset.originalId;
-            if (originalId) {
-                subDomain.push(originalId);
-            }
-
-            domain.push(['original_id', 'in', subDomain]);
```
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/static/src/components/media_dialog/image_selector.js:153]: If things remain as they are, it won't be necessary after the commit introducing the media model, as optimized copies are not media records, so they are not fetched.
But if we stop at this commit, you're absolutely right. I'll change that.

File: addons/test_website/tests/test_media.py [hunk 9]
```diff
@@ -1,8 +1,6 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import base64
```
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/test_website/tests/test_media.py:4]: By default, the `media` model is only linked to an `ir.attachment` through `fields.Binary()`, which means it only creates an attachment if you pass it a binary content. But the image dialog allows to add an image through a URL. In that case, the creation of the attachment through the media model is done explicitly.
By "true res_model and res_id", I mean the `res_model` and `res_id` of the page/record in which the media is inserted (typically on a website page: `ir.ui.view` with an overridden ID of `0`), as opposed to the `res_model` & `res_id` of the media record which are stored in the `ir.attachment` record. (It's not specifically linked to the URL media type, it's for all media records.)

Does that make it clearer? :thinking: 
I can rephrase it to be clearer.

File: addons/web_editor/controllers/main.py [hunk 10]
```diff
@@ -63,8 +63,13 @@ def get_existing_attachment(IrAttachment, vals):
     so, None otherwise.
     """
     fields = dict(vals)
-    # Falsy res_id defaults to 0 on attachment creation.
-    fields['res_id'] = fields.get('res_id') or 0
+    if fields['res_model'] == 'web_editor.media':
+        # For a media attachment, the res_id is the media's id, which (if it
+        # exists) we don't know at this point. We do the comparison on the rest.
+        fields.pop('res_id', None)
```
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/controllers/main.py:69]: `get_existing_attachment` checks the attachment itself, not the media (as we're in the middle of trying to upload an image, we only have the data of the file itself). So the `res_id` that is passed here as a value is the id of the current record in which the attachment is inserted.
But as said earlier, the "true" `res_id` is stored on the media record, not the attachment anymore. So if I don't do this here, it will never find any existing attachment, because as long as I don't know if an attachment exists, I can't find the related media record, so I can't access the `res_id` that is checked here.

I was about to say that "if I can get rid of all routes that create basic attachments in `web_editor`, that condition wouldn't have to be an exception anymore and I could just remove the `res_id` from the comparison", but the function is at least used in `modify_image()` which doesn't create media records (and from my understanding I don't think it should?).

File: addons/web_editor/controllers/main.py (hunk 6 above)
Comment (robinlej) [reply addons/web_editor/controllers/main.py:312]: :eyes: 
I've got to test things to answer here. I remember that things weren't working out of the box, but I don't remember the specifics and why I didn't implement it there.

File: addons/web_editor/models/media.py (hunk 7 above)
Comment (robinlej) [reply addons/web_editor/models/media.py:129]: Same reason as the other explanations: `res_id` and `res_model`. I need the actual `res_id` / `res_model` (on the media record) but also a bunch of info on the attachment for the media dialog to treat the info. So I need a join on `ir.attachment` with a combined `where` clause.
JS-side, that would amount to at least 2 non-optimized requests:
- either I start by calling `ir.attachment`, but it cannot be optimized because how do I respect the limit while I don't know which attachments I should fetch for the current record?
- or I start with `web_editor.media`, but I don't have all the discriminating info, specifically the name and mimetype. So I end up with the same problem.

Comment (robinlej) [reply addons/web_editor/models/media.py:129]: Note that if I store an explicit relationship to the `attachment_id` as a field of `media`, I think there might be ways to simplify things with the ORM and maybe even manage it in one go from the JS (didn't test nor dug in that direction though, so I can't confirm for now).

File: addons/web_editor/controllers/main.py (hunk 6 above)
Comment (robinlej) [reply addons/web_editor/controllers/main.py:312]: > Media dialog "Files" -> Don't you think this should use the new web_editor.media model too? Maybe with a field type to distinguish Images, Files and maybe later Videos ?

I'm not sure the `type` field is needed, we can always check the mimetype on the `attachment`.
But here, files through the media dialog _are_ saved media, that's not the problem.
That one route is in knowledge, with a use spanning between knowledge and modules with an html_field, and it triggers a `macro` which I can't wrap my head around:
1) the attachment / media exists in a knowledge article
2) by clicking on a button, it is created and linked to the other module's view
3) it is used in the chatter => that's the part that fails, because it's not looking for a media but for an attachment, with the `res_id` / `res_model` that are now stored on the media. Except I haven't found yet where that happens, and I'm not convinced it should be anything else than an attachment anyway.

> Logo upload during website configurator flow, can't it be also a new web_editor.media record. Really, the logo should always be suggested in the media dialog whatever page or record you are.

The `website.logo` is already a binary field itself. At first glance, it looks like I can modify the route, but the logo won't appear in the media dialog anyway (it's already not the case for now), because upon save, the attachment's `res_model` / `res_id` are overridden with `website.logo`. Which means that, by this point, there is a zombie `web_editor.media` not attached to any attachment anymore (I guess I could add some sort of autovacuum?).

Maybe there is also a test somewhere that doesn't like it, but I'm not sure as I can't find any by just checking the code. I guess that if I do it, I'll see when I push next time. ^^

File: addons/test_website/tests/test_media.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/test_website/tests/test_media.py:4]: It does make it clearer to mention the "image added through an URL", that's what I missed :+1:

File: addons/web_editor/controllers/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/controllers/main.py:69]: Ok I see. Then what I don't see is the use we have for `get_existing_attachment` at all when `fields['res_model'] == 'web_editor.media'`. Maybe it should not be called at all? :open_mouth:

Maybe it is to prevent uploading the same image twice in the media dialog for the same model+id+field? :open_mouth: Did not know we had that sort of check... do we? If this is it: is not this update breaking something? If you prevent creation of duplicate for all related-to-media attachments without considering what the web_editor.media are themselves attached to? Like you could not upload the same image on different products maybe?

=> In any case, there is something strange here.

File: addons/web_editor/controllers/main.py (hunk 6 above)
Comment (qsm-odoo) [reply addons/web_editor/controllers/main.py:312]: > But here, files through the media dialog are saved media, that's not the problem.
> That one route is in knowledge,

Oh, yes, I probably saw "File" and supposed it was the documents section of the media dialog without seeing I was in knowledge.

> with a use spanning between knowledge and modules with an html_field, and it triggers a macro which I can't wrap my head around

I let you investigate, if really you don't find, we'll talk about it IRL. In any case, this route should not exist anymore, at least not in web_editor.

> but the logo won't appear in the media dialog anyway (it's already not the case for now)

That's something that should change both for the sake of this task and from a functional point of view.
1. Maybe we should simply open the media dialog when asking the user for its logo. Debatable, probably useless from a functional point of view but who knows.
2. In any case, what should happen after the user choice:
  - A web_editor.media record is created for the logo. Its res_model+res_id+res_field can all be empty IMO. Such records with empty values for res_XXX should always be listed. Like all res_model=products + res_id=None ones should be listed on all products IMO. Like all res_model=None + res_field=something should be listed when editor a field named something IMO. Etc.
  - An ir.attachment should be created (automatically I guess?) to be linked to the website.logo field. Its original_id should be the web_editor.media.

File: addons/web_editor/controllers/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/controllers/main.py:69]: > Maybe it is to prevent uploading the same image twice in the media dialog for the same model+id+field?

Yes that's the point, and yes you were right, that created some bugs (the same image could be used on 2 different models, but then couldn't be retrieved on the 2nd one). I'll add a check about that.

File: addons/web_editor/models/media.py (hunk 7 above)
Comment (robinlej) [reply addons/web_editor/models/media.py]: > That's what should be investigated before resorting to writing SQL queries

I confirm that there's no need for a SQL query if I store `media.attachment_id` and make a call with `.webSearchRead()`.

> Ideally, the media dialog should not know the ir.attachment model exists.

> The mimetype: should just not be needed. Why is it? 

That's not the only one, here are all the things that are stored on the attachment (not yet on the media) and retrieved for the dialog:

```js
'description',
'mimetype',
'checksum',
'type',
'access_token',
'original_id',
'image_src',
'image_width',
'image_height',
```

Some are then saved as data-attributes on the images, and then used in many different places in the codebase (both html_editor/web_editor & website). Either removing them from the media info (mimetype) or modifying their meaning (original_id as suggested in another of your comments) would have a lot of impacts. Are we sure that's what we want?

In addition, while some may be transferred to the media, a few are tied to the attachment itself: namely checksum, access_token, image_src / image_width / image_height (those 3 are computed, but image_src in particular is used in some places where a `media` record wouldn't be expected). We can discuss it, but I don't think we can do without them, meaning the subquery to `ir.attachment` needs to happen anyway?

Comment (qsm-odoo) [reply addons/web_editor/models/media.py]: > That's not the only one

Problem is that I'll just ask the question for the rest of them then ^^ Without looking at the code, I really don't understand why the media dialog would want to know about any of those fields :grimacing: 

> Some are then saved as data-attributes on the images

Just the mimetype and the original_id no? But that won't be true anymore after LOCO's task... :grimacing: 

Probably worth speaking about it IRL, it will be easier (remind me maybe Wednesday?)
